### PR TITLE
feat: better inline images

### PR DESCRIPTION
**Pull request by marbemac**
Part of https://github.com/stoplightio/mosaic/issues/219

Basically does two main things: 

1. Add "inline" prop to images that should be rendered inline by downstream renderers.
2. Adjusts the unwrap images plugin to not unwrap if there is more than one image/link in a single paragraph.

### CODE REVIEW

File: src/plugins/run/__tests__/__fixtures__/images/basic-inline.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+Image in a paragraph: ![should have inline][ecologi_image].
```
Comment (marbemac) [inline src/plugins/run/__tests__/__fixtures__/images/basic-inline.md:1]: This markdown fixture captures the meat of the change in this PR. Basically a plugin to add "inline=true" to images that we shouldn't treat as block level elements in downstream renderers (like markdown-viewer).

File: src/plugins/run/__tests__/__fixtures__/images/unwrap.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
+Image in a paragraph: ![should NOT unwrap][ecologi_image].
+
+![should unwrap][ecologi_image]
+
+[![should unwrap][ecologi_image]][ecologi]
+
+![should NOT unwrap][ecologi_image] [![should NOT unwrap][ecologi_image]][ecologi]
```
Comment (marbemac) [inline src/plugins/run/__tests__/__fixtures__/images/unwrap.md:7]: This is the other change - don't unwrap a paragraph of images/links if there is more than one image/link inside of it (helps support github readme badge link images that are rendered inline in a single paragraph, etc).

File: src/reader/__tests__/reader.spec.ts [hunk 3]
```diff
@@ -270,137 +270,138 @@ var x = true;
       const mdastTree = mdReader.fromLang(input);
 
       expect(prettyStringify(mdastTree)).toMatchInlineSnapshot(`
-        "<>
```
Comment (marbemac) [inline src/reader/__tests__/reader.spec.ts:273]: Lol, this snapshot decided to change indentation - nothing inside of it actually changed.

Comment (P0lip) [reply src/reader/__tests__/reader.spec.ts:273]: Isn't it because of the dep bump?

Comment (marbemac) [reply src/reader/__tests__/reader.spec.ts:273]: Yes that's quite likely

File: src/plugins/run/unwrap-images.ts [hunk 4]
```diff
@@ -28,11 +28,15 @@ function applicable(node: Parent, inLink?: boolean): boolean | null {
...
+  let nonwhitespaceChildren = 0;
 
   while (++index < length) {
     child = children[index];
 
-    if (whitespace(child)) {
+    const isWhitespace = whitespace(child);
+    if (!isWhitespace) nonwhitespaceChildren++;
+
+    if (isWhitespace) {
```
Comment (P0lip) [inline src/plugins/run/unwrap-images.ts]: if/else not good anymore?

File: src/plugins/run/unwrap-images.ts [hunk 5]
```diff
@@ -51,5 +55,7 @@ function applicable(node: Parent, inLink?: boolean): boolean | null {
     }
   }
 
+  if (nonwhitespaceChildren > 1) return false;
```
Comment (P0lip) [inline src/plugins/run/unwrap-images.ts]: seems like we could move it to the loop's body to abort the iteration early?

File: src/plugins/run/annotations.ts [hunk 6]
```diff
@@ -189,10 +189,29 @@ function processNode(node: MDAST.Content, annotations?: object): MDAST.Content {
...
+        hProperties: normalizeAnnotationsForHast(annotations),
       },
     };
   }
 
   return node;
 }
+
+// micromark ecosystem passes data through html, and then to react
+// HTML does not allow for boolean properties, so here we stringify boolean values so that they can pass through
```
Comment (P0lip) [inline src/plugins/run/annotations.ts:201]: > HTML does not allow for boolean properties

How so?
There are boolean attributes.
https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attributes

File: src/plugins/run/code.ts [hunk 7]
```diff
@@ -156,6 +156,7 @@ function handleLegacyAnnotations(annotations: MDAST.Code['annotations']) {
     // @ts-expect-error type is no longer part of the typings, it is deprecated
     const type = annotations.type;
     if (type === 'json_schema') {
+      // @ts-expect-error
```
Comment (P0lip) [inline src/plugins/run/code.ts:159]: Can't we just change the value of `jsonSchema` to true?

**Review by P0lip — COMMENTED** [review]
Left a few comments, but looks good overall

File: src/plugins/run/unwrap-images.ts (hunk 4 above)
Comment (marbemac) [reply src/plugins/run/unwrap-images.ts]: `nonwhitespaceChildren++` needs to occur in all of the "else" statements below 39 (line 42, line 44, line 53) - the behavior would change if I adjusted to make this if/else

File: src/plugins/run/unwrap-images.ts (hunk 5 above)
Comment (marbemac) [reply src/plugins/run/unwrap-images.ts]: yes good call

File: src/plugins/run/annotations.ts (hunk 6 above)
Comment (marbemac) [reply src/plugins/run/annotations.ts:201]: I meant no boolean attributes like `disabled=true`. The html boolean attributes basically translate to `disabled=""`, and as far as I'm aware there is no way to pass false, so really it's a "true" attribute (not boolean). Currently in `markdown-viewer` it's checking if these boolean attributes do or do not equal `""`, which just seemed more confusing that checking against `"true"` or `"false"`. But tbh I don't mind a whole lot either way, just thought that explicit "true" / "false" was more clear than "".

File: src/plugins/run/code.ts (hunk 7 above)
Comment (marbemac) [reply src/plugins/run/code.ts:159]: Yeah but if we do that then the output in the HTML is `jsonSchema`, which ultimately translates to `jsonSchema=""`. Per above comment this just seemed a bit confusing to me (versus explicit "true"/"false" values), but I can change it back if you feel strongly about it no problem just LMK.

(note that this is HTML, not JSX)

Screen Shot 2021-06-17 at 12 57 03 PM

File: src/plugins/run/annotations.ts (hunk 6 above)
Comment (P0lip) [reply src/plugins/run/annotations.ts:201]: > and as far as I'm aware there is no way to pass false, so really it's a "true" attribute (not boolean)

yeah, as the spec states:

> The presence of a boolean attribute on an element represents the true value, and the absence of the attribute represents the false value.

Comment (P0lip) [reply src/plugins/run/annotations.ts:201]: > But tbh I don't mind a whole lot either way, just thought that explicit "true" / "false" was more clear than "".

It's actually less clear IMHO. We should try to stick to HTML, I think.

**Review by P0lip — APPROVED** [review]
In general :ok_hand: though I do prefer us to stick to HTML instead of relying on 'true' | 'false' which is not that standard IMHO

**Comment by marbemac** [comment]
> In general :ok_hand: though I do prefer us to stick to HTML instead of relying on 'true' | 'false' which is not that standard IMHO

It's less expressive though - basically have 2 states (true and false) rather than three states (true, false, and undefined). Specifically thinking for https://github.com/stoplightio/markdown-viewer/pull/96/files?file-filters%5B%5D=.tsx#diff-889173fec0fc121845ec5f8022bb36be8ad8f0b1f354ed87ce7e0b2ee4f35ad2R75

Basically we want to allow the end user to opt out of the styling we apply to images in general, and I figured explicitly passing "false" could do the trick. I guess could do "" is "true", "false" is "false", and absence of value is undefined - what do you think?

**Comment by P0lip** [comment]
> Basically we want to allow the end user to opt out of the styling we apply to images in general, and I figured explicitly passing "false" could do the trick. I guess could do "" is "true", "false" is "false", and absence of value is undefined - what do you think?

`null` could be better.
Regardless of that, I think we can merge this PR.

**Pull request merged**